Variant inference in case expressions

GearsDatapacks opened the pull request:
Closes #2060
This PR allows the compiler to infer the variant of values when pattern matching. This allows you to perform safe record updates when the variant is known, and access fields which aren't common to all variants

inoas commented:
@GearsDatapacks wow these two changes coming from you are worth to use nightly all the time ASAP! Thank you so much!

lpil commented on compiler-core/src/type_.rs:
```diff
@@ -63,6 +63,7 @@ pub enum Type {
         module: EcoString,
         name: EcoString,
         args: Vec<Arc<Type>>,
+        constructor_index: Option<u16>,
```
What's this field? It's not clear to me how a type has a constructor index.

GearsDatapacks replied:
This represents which variant a certain value has when the variant is known, so for example:
```gleam
type Wibble { Wibble Wobble }

fn wibble(w: Wibble) {
  //      ^ We have no idea which variant this is, so constructor_index is None
  case w {
    Wibble as w -> todo
    //        ^ We know that this is the Wibble variant, so constructor_index would be Some(0)
    Wobble as w -> todo
  }
}
```

This perhaps isn't the best way to implement this, but it's the way I came up with that made fairly minimal changes to the way that the type system is represented internally. If you have another idea, I'm open to suggestions

lpil replied:
I don't have any particular suggestions for the design, I think this is quite clever.

I'm wondering if there's a clearer name that we can use here as it wasn't obvious to me what this means. At the least it needs a good deal of documentation.

lpil commented on compiler-core/src/type_.rs:
```diff
@@ -1142,6 +1195,7 @@ pub fn generalise(t: Arc<Type>) -> Arc<Type> {
                 package: package.clone(),
                 name: name.clone(),
                 args,
+                constructor_index: *constructor_index,
```
I think this may be a bug? Shouldn't generalising a type discard what variant it is?

GearsDatapacks replied:
Oops, I must have missed that one

GearsDatapacks commented:
Hopefully the functionality of the `constructor_index` field should be clearer now

GearsDatapacks commented on compiler-core/src/type_/tests/errors.rs:
```diff
@@ -2434,8 +2434,8 @@ pub type Wibble {
 
-pub fn main() {
-  Wibble(1).field
+pub fn main(wibble: Wibble) {
+  wibble.field
```
See above

GearsDatapacks commented on compiler-core/src/type_/tests/errors.rs:
```diff
@@ -2471,3 +2471,38 @@ pub fn main() {
+}
+"#
+    );
+}
```
I should be able to make that compile, although it doesn't currently. Either way, I'll add a test later

GearsDatapacks commented on compiler-core/src/type_/tests/snapshots/gleam_core__type___tests__errors__record_update_unknown_variant.snap:
```diff
@@ -0,0 +1,16 @@
+source: compiler-core/src/type_/tests/errors.rs
+expression: "\npub type Wibble {\n  Wibble(wibble: Int, wubble: Bool)\n  Wobble(wobble: Int, wubble: Bool)\n}\n\npub fn wibble(value: Wibble) {\n  Wibble(..value, wubble: True)\n}\n"
+---
+error: Unsafe record update
+  ┌─ /src/one/two.gleam:8:3
+  │
+8 │   Wibble(..value, wubble: True)
+  │   ^^^^^^ I can't tell this is always the right constructor
+
+This type has multiple constructors so it cannot be safely updated. If
+this value was one of the other variants then the update would be produce
+incorrect results.
+
+Consider pattern matching on it with a case expression and then
+constructing a new record with its values.
```
This error message hasn't changed in this PR; it's always been like that, but now is probably a good time to improve it, thanks!

GearsDatapacks commented on compiler-core/src/type_/tests/snapshots/gleam_core__type___tests__errors__record_update_wrong_variant.snap:
```diff
@@ -0,0 +1,14 @@
+source: compiler-core/src/type_/tests/errors.rs
+expression: "\npub type MyRecord {\n  A(common: Int, other: String)\n  B(common: Int, different: Float)\n}\n\npub fn b_to_a(value: MyRecord) {\n  case value {\n    A(..) -> value\n    B(..) as b -> A(..b, other: \"Hi\")\n  }\n}\n"
+---
+error: Mismatched record update
+   ┌─ /src/one/two.gleam:10:19
+   │
+10 │     B(..) as b -> A(..b, other: "Hi")
+   │                   ^ This is never the right constructor
+
+This value is never the same constructor as the one being updated.
+
+If you want to change one variant of a type into another, each field of the
+new type must be specified explicitly.
```
Good idea!

GearsDatapacks commented:
Thanks for the review Jak! I'll get to those points as soon as I have time

giacomocavalieri commented:
Playing around with it I found two small bugs.
```gleam
pub fn main() {
  let a = todo

  case a {
    Wibble(..) -> Wibble(..a, b: 1)
//                ^^^^^^^^^^^^^^^^^ No complaints here
    Wobble(..) -> Wobble(..a, c: 1)
//                ^^^^^^ Mismatched record update
  }
}
```
I'd expect the compiler to complain on both, instead here it looks like it's narrowing the `todo` to be a `Wibble`.

And the second one is:
```gleam
pub type Wibble {
  Wibble(a: Int, b: Int)
  Wobble(a: Int, c: String)
}

pub fn main() {
  let a = case todo {
    Wibble(..) as b -> Wibble(..b, b: 1)
    Wobble(..) as b -> Wobble(..b, c: "a")
  }

  a.b
//^^^ This shouldn't compile but it does
}
```

lpil commented:
@GearsDatapacks Note I renamed some stuff to make naming a bit more consistent as `main` currently is inconsistent

GearsDatapacks commented:
I've addressed all the points in your review now. This now compiles:
```gleam
case some_wibble {
  Wibble(..) -> some_wibble.field_only_present_on_wibble
  _ -> todo
}
```

Both bugs you found are now also fixed.
I just made it so whenever a type with a narrowed variant is unified with an unbound type variable, it discards the narrowed variant. That seemed to solve all the potentially problems that I could think of. It does very slightly reduce the amount which we can infer about narrowed type variants, but it was the only way I could think of to fix these problems without making big changes to the unification process.

GearsDatapacks commented:
I also have a question about an error message which could potentially be confusing to the user. For example:
```gleam
type Wibble {
  Wibble(a: Int, b: Int, c: Int)
  Wobble(a: Int, d: Int)
}

pub fn a(wibble: Wibble) {
  wibble.wobble // Error: The type `Wibble` only has the `a` field
}

pub fn b() {
  let wibble = Wibble(a: 1, b: 2, c: 3)
  wibble.wobble // Error: The type `Wibble` only has the fields `a`, `b` and `c`
}
```

In the second case the compiler has inferred that `wibble` is of the `Wibble` variant, so it says that it has all three of those fields. However, this extra information isn't reflected anywhere in the error message. Should we print that somehow?

lpil commented:
Nice work!

I've found a bug where the narrowing is escaping out of the case expression. This invalid code compiles and crashes at runtime

```gleam
import gleam/io

pub type Thingy {
  A(a: Int)
  B(x: Int, b: Int)
}

pub fn main() {
  A(1) |> fun |> io.debug
}

fn fun(x) {
  case x {
    A(..) -> x.a
    B(..) -> x.b
  }
  x.b
}
```

GearsDatapacks commented:
Interesting, I wonder why that's happening. I'll look into it, thanks!

GearsDatapacks commented:
Found the issue. It was to do with `TypeVar::Link`, which allows shared mutability with `Arc<RefCell<Type>>`. This meant that when I was changing the `narrowed_variant` field of a type in the current scope, it was also mutating the type in the parent scope as they were linked together. To prevent this, I just called `collapse_links` when creating the type in the inner scope.

lpil commented:
Hello! I'm back from CodeBEAM now.

> It was to do with TypeVar::Link, which allows shared mutability with Arc<RefCell<Type>>.

This is intentional, the mutability of types is used to propagate inferred type information to linked type variables. We'll need to carefully test this to ensure that types are still inferred correctly.

I'll try and review this soon.

giacomocavalieri commented on CHANGELOG.md:
```diff
@@ -109,6 +109,25 @@
   the Erlang target if any of their arguments were discarded.
   ([Giacomo Cavalieri](https://github.com/giacomocavalieri))
 
+- The compiler now narrows the variant of custom types when pattern matching,
+  allowing for safe record updates if the variant is known:
```
Talking with Louis and Hayleigh at code beam we agreed that we should be really clear in framing this as "safe record updates" instead of _narrowing_ of any type to avoid misleading people into thinking this is type narrowing. This could be reworded as:
```suggestion
- The compiler can now allow you to perform record updates for specific
  variants of a type after pattern matching on it:
```

GearsDatapacks replied:
Sounds good!

GearsDatapacks commented:
I've added a couple of tests to make sure this change doesn't break type inference, but I honestly couldn't think of that many cases to test. Anyone who is more familiar with the type inference system, please suggest more cases if you come up with any

lpil commented on compiler-core/src/error.rs:
```diff
@@ -1881,31 +1882,57 @@ assigned variables to all of them."
+                    match reason {
+                        UnsafeRecordUpdateReason::UnknownVariant {constructed_variant} => {
+                            let text = wrap_format!("
+I cannot use this value to build an updated `{constructed_variant}` as it could be
```
Remove this hard-coded newline by adding back the \ at the end please

GearsDatapacks replied:
Done!

lpil reviewed (commented):
This looks super nice! I've been testing it and it seems to work great!

One tiny note inline. It would be great if other folks could test this out and try and find any bugs, in case Gears and I have missed anything. @giacomocavalieri says he will be able to test tomorrow or so.

💜

GearsDatapacks commented:
I have now implemented exhaustiveness checking for narrowed types, as discussed on Discord.
This allows stuff like:
```gleam
type Wibble {
  Wibble(Bool)
  Wobble(Int)
}

pub fn main(b: Bool) {
  let always_wibble = Wibble(b)
  case always_wibble {
    Wibble(True) -> 1
    Wibble(False) -> 0
  }
  // ^ This case expression is exhaustive!
}
```

GearsDatapacks commented:
I should add version tracking to this as well. I was thinking about it before, and wasn't really sure if there was a good way to do it, but it should actually be pretty simple.

giacomocavalieri commented:
With some type system changes (like inference of anonymous functions in pipes) we decided to exclude those from the tracking. Here it seems like it wouldn't be too hard to check if someone is relying on narrowing when updating/accessing fields so I think it would be nice to track it

GearsDatapacks commented:
Yep. I'm not sure I'll be able to track the exhaustiveness checking for narrowed types, but record access and record updates should be doable

giacomocavalieri commented on compiler-core/src/type_.rs:
```diff
@@ -1279,6 +1279,9 @@ pub fn generalise(t: Arc<Type>) -> Arc<Type> {
 pub enum FieldAccessUsage {
     /// Used as `thing.field()`
     MethodCall,
+    /// Used internally when trying to access a field when
+    /// performing record updates
+    RecordUpdate,
```
It's not clear to me what this means, just by the name and description I'd guess it's this
```gleam
Wobble(..wobble, field: other.field)
//                            ^^^^^
```
Did you mean access to a field that is inferred to be safe after narrowing? I'd change the name in that case

GearsDatapacks replied:
Yeah sorry I'll clarify that

giacomocavalieri commented on compiler-core/src/type_/error.rs:
```diff
@@ -841,6 +841,7 @@ pub enum FeatureKind {
     NestedTupleAccess,
     InternalAnnotation,
     AtInJavascriptModules,
+    TypeNarrowing,
```
I'd define two different variants to produce nicer messages:
- one to talk about narrowing for safe record updates
- one to talk about narrowing for safe accessors

GearsDatapacks replied:
Good idea!

lpil reviewed (commented):
This is looking super nice! I'm trying to give this a thorough test at the moment as I'm nervous about the change to type mutation. I'd appreciate others doing the same and using this branch for lots of programming.

GearsDatapacks commented on compiler-core/src/type_/environment.rs:
```diff
@@ -812,7 +835,11 @@ pub fn unify(t1: Arc<Type>, t2: Arc<Type>) -> Result<(), UnifyError> {
+                t2.generalise_custom_type_variant();
+                *type_.borrow_mut() = TypeVar::Link {
+                    type_: Arc::new(t2),
+                };
```
Why would it? All the change does is create a new `Arc`, instead of cloning the existing `t2` one.

lpil replied:
Because it no longer mutates, which is how I made sure that linked types are updated together when there's more context.

GearsDatapacks commented:
All your suggested wording changes have been implemented

giacomocavalieri commented on CHANGELOG.md:
```diff
@@ -126,6 +126,26 @@
 - Optimised creation of bit arrays on the JavaScript target.
   ([Richard Viney](https://github.com/richard-viney))
 
+- The compiler can now infer the variant of custom types when pattern matching,
+  allowing for safe record updates if the variant is known, or access fields
+  which are only present on one variant:
```
Instead of cramming it into a single entry I'd split it in two: one for the record updates and one for the record access, so the example can be more focused

GearsDatapacks replied:
Good idea, I wasn't super happy with how the entry looked

lpil replied:
It'd also need to bring up pattern matching!

lpil reviewed (approved):
Thank you!!
